plutus-wallet-api/ledger/Ledger/Constraints/OnChain.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE MonoLocalBinds     #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeApplications   #-}
+{-# OPTIONS_GHC -Wno-simplifiable-class-constraints #-}
+{-# OPTIONS_GHC -fno-strictness #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+module Ledger.Constraints.OnChain where
+
+import qualified Data.Aeson                       as Aeson
+import           Data.Text.Prettyprint.Doc
+import           GHC.Generics                     (Generic)
+
+import           Language.PlutusTx                (IsData (..))
+import qualified Language.PlutusTx                as PlutusTx
+import           Language.PlutusTx.Prelude
+
+import           Ledger.Constraints.TxConstraints
+import           Ledger.Constraints.UtxoLike
+import           Ledger.Crypto                    (PubKey)
+import           Ledger.Interval                  (contains)
+import           Ledger.Scripts                   (DataValue (..), ValidatorHash)
+import           Ledger.Tx                        (TxOutType (..))
+import           Ledger.Validation                (PendingTx, PendingTx' (..), TxOut (..))
+import qualified Ledger.Validation                as V
+import           Ledger.Value                     (Value, leq)
+
+import qualified Prelude                          as Haskell
+
+-- TODO: Use something other than unit for the input constraints (need
+-- to think about representation)
```

Comment:
Yes. The problem is when we go from `PendingTxConstraints` to `LedgerConstraints`, we need to resolve all `ValidatorHash` values from the on-chain inputs to `Validator`s that we can attach to the transaction. I couldn't think of a nice way to do that, other than requiring the user to pass in a `ValidatorHash -> Maybe Validator`. Maybe there's a more general solution, a mechanism for adding proofs to transactions (signatures, validators, data values, etc.) in the client code. But I don't know what that would look like.
 
Making the transition from on-chain to off-chain `TxConstraints` as easy as possible was an important goal for this whole system, because that's the only way we can avoid implementing most of the same logic twice (like we had to do with `payments` and `transitions` before). That's why the constraints are rather specific: They contain enough information to turn the `PendingTxConstraints` into a transaction that can be submitted, after balancing and signing.

Reply:
Hm, why not do it the other way around? i.e. have things produce `LedgerConstraints` and then throw away details to get to `PendingTxConstraints`.

Reply:
I tried that but functions that produce `LedgerConstraints` usually involve hashing the validator scripts somewhere, so you can't compile them with the plugin.

Reply:
On reflection, perhaps passing the `ValidatorHash -> Maybe Validator` function isn't so bad after all. Consider that in the real transaction body (and probably in our representation thereof) there really are only hashes of things (scripts, data) and those are resolved by a hashmap on the side.

So we could have the constraint-generating functions just use the hashes. They'll also need to have them passed in, since that's how you get the hashes of scripts without computing them on chain. Then the on-chain one gets passed the hashes when we construct the validator, and the off-chain one similarly, but we also use that moment to create the hashmap that needs to go along with it for constructing the final transaction.

Does that make sense?